test/e2e/framework/service/jig.go

```diff
@@ -697,6 +757,108 @@ func (j *TestJig) waitForPodsReady(namespace string, pods []string) error {
+	cmd := ""
+	switch protocol {
+	case v1.ProtocolTCP:
+		cmd = fmt.Sprintf("nc -z -t -w 2 %s %v", endpoint, port)
+	case v1.ProtocolUDP:
+		cmd = fmt.Sprintf("nc -z -u -w 2 %s %v", endpoint, port)
+	default:
+		e2elog.Failf("Service reachablity check is not supported for %v", protocol)
+	}
+	if cmd != "" {
+		_, err := framework.RunHostCmd(execPod.Namespace, execPod.Name, cmd)
+		framework.ExpectNoError(err, "Service is not reachable on following endpoint %s over %s protocol", ep, protocol)
+	}
+}
+
+// checkClusterIPServiceReachability ensures that service of type ClusterIP is reachable over
+// - ServiceName:ServicePort, ClusterIP:ServicePort
+func (j *TestJig) checkClusterIPServiceReachability(namespace string, svc *v1.Service, pod *v1.Pod) {
+	clusterIP := svc.Spec.ClusterIP
+	servicePorts := svc.Spec.Ports
+
+	j.WaitForAvailableEndpoint(namespace, svc.Name, ServiceEndpointsTimeout)
+
+	for _, servicePort := range servicePorts {
+		testReachabilityOverServiceName(svc.Name, servicePort, pod)
+		testReachabilityOverClusterIP(clusterIP, servicePort, pod)
+	}
+}
+
+// checkNodePortServiceReachability ensures that service of type nodePort are reachable
+// - Internal clients should be reachable to service over -
+//   	ServiceName:ServicePort, ClusterIP:ServicePort and NodeInternalIPs:NodePort
+// - External clients should be reachable to service over -
+//   	NodePublicIPs:NodePort
+func (j *TestJig) checkNodePortServiceReachability(namespace string, svc *v1.Service, pod *v1.Pod) {
+	clusterIP := svc.Spec.ClusterIP
+	servicePorts := svc.Spec.Ports
+
+	// Consider only 2 nodes for testing
+	nodes := j.GetNodes(2)
+
+	j.WaitForAvailableEndpoint(namespace, svc.Name, ServiceEndpointsTimeout)
+
+	for _, servicePort := range servicePorts {
+		testReachabilityOverServiceName(svc.Name, servicePort, pod)
+		testReachabilityOverClusterIP(clusterIP, servicePort, pod)
+		testReachabilityOverNodePorts(nodes, servicePort, pod)
+	}
+}
+
+// checkExternalServiceReachability ensures service of type externalName resolves to IP address and no fake externalName is set
+// FQDN of kubernetes is used as externalName(for air tight platforms).
+func (j *TestJig) checkExternalServiceReachability(svc *v1.Service, pod *v1.Pod) {
+	// Service must resolve to IP
+	cmd := fmt.Sprintf("nslookup %s", svc.Name)
+	_, err := framework.RunHostCmd(pod.Namespace, pod.Name, cmd)
+	framework.ExpectNoError(err, "ExternalName service must resolve to IP")
+}
+
+// CheckServiceReachability ensures that request are served by the services.
+func (j *TestJig) CheckServiceReachability(namespace string, svc *v1.Service, pod *v1.Pod) {
+	svcType := svc.Spec.Type
+
+	j.SanityCheckService(svc, svcType)
+
+	switch svcType {
+	case v1.ServiceTypeClusterIP:
+		j.checkClusterIPServiceReachability(namespace, svc, pod)
+	case v1.ServiceTypeNodePort:
+		j.checkNodePortServiceReachability(namespace, svc, pod)
+	case v1.ServiceTypeExternalName:
+		j.checkExternalServiceReachability(svc, pod)
+	default:
```

Comment:
Similar to previous comment, this would also fail silently for type=LoadBalancer

Reply:
Done.

Reply:
Comment in CheckServiceReachability: "Only supports Services with type ClusterIP, NodePort and ExternalName"

Reply:
Comment in CheckServiceReachability: "Only supports Services with type ClusterIP, NodePort and ExternalName"

Reply:
Done.